system/lib/emmalloc.c

```diff
@@ -91,8 +91,16 @@ static_assert(alignof(max_align_t) == 8, "max_align_t must be correct");
 typedef struct Region
 {
   uint32_t size;
-  // Use a circular doubly linked list to represent free region data.
-  struct Region *prev, *next;
+  union {
+    // Use a circular doubly linked list to represent free region data.
+    struct FreeRegionPayload {
+      struct Region *prev, *next;
+    } freeInfo;
+    struct RootRegionPayload {
+      struct Region *next;
+      uint8_t* endPtr;
+    } rootInfo;
+  };
```

Comment:
After this change the code is littered with these `->freeInfo.` and `->rootInfo.` accesses - I don't think that particularly helps with readability.

Agree that the bare pointer casting becomes clearer though.

How about instead creating a second struct
```c++
struct RootBlock
{
  uint32_t size;
  struct Region *nextRootBlock;
  uint8_t *regionBlockEndPtr;
};
```
and casting to that where necessary - that way the `->freeInfo.` dereferences won't occur?

Reply:
I gave it go.  It seems a little less type safe as I end up needing to use explict casts when calling any of the helper functions that operate on both region types.   WDYT: #17685.   I think its slightly less type safe but but I could go either way.